# Camada Bronze

Ingestão e visualização dos dados

In [0]:
%fs ls /FileStore/tables/boston_marathon_data/

path name size modificationTime dbfs:/FileStore/tables/boston_marathon_data/boston_marathon_2023.csv boston_marathon_2023.csv 3169826 1744291116000 dbfs:/FileStore/tables/boston_marathon_data/boston_marathon_2024.csv boston_marathon_2024.csv 4841636 1744291117000

In [0]:
#Ingestão de arquivos por Pyspark - Dados de 2024

df_2024 = spark.read.csv("/FileStore/tables/boston_marathon_data/boston_marathon_2024.csv",header = True)
display(df_2024.limit(5))

BibNumber FullName SortName AgeOnRaceDay Gender City StateAbbrev StateName Zip CountryOfResAbbrev CountryOfResName CountryOfCtzAbbrev CountryOfCtzName OfficialTime RankOverall RankOverGender RankOverDivision EventGroup SubGroupLabel SubGroup 2 Sisay Lemma Lemma, Sisay 33 M Addis Ababa null null null ETH Ethiopia ETH Ethiopia 2:06:17 1 1 1 Runners null null 8 Mohamed Esa Esa, Mohamed 23 M Addis Ababa null null null ETH Ethiopia ETH Ethiopia 2:06:58 2 2 2 Runners null null 1 Evans Chebet Chebet, Evans 35 M Kapsabet null null null KEN Kenya KEN Kenya 2:07:22 3 3 3 Runners null null 7 John Korir Korir, John 27 M Eldoret null null null KEN Kenya KEN Kenya 2:07:40 4 4 4 Runners null null 14 Albert Korir Korir, Albert 30 M Nairobi null null null KEN Kenya KEN Kenya 2:07:47 5 5 5 Runners null null

In [0]:
#Ingestão de arquivos por Pyspark - Dados de 2023

df_2023 = spark.read.csv("/FileStore/tables/boston_marathon_data/boston_marathon_2023.csv",header = True)
display(df_2023.limit(5))

age_group place_overall place_gender place_division name team bib_number half_time finish_net finish_gun gender half_time_sec finish_net_sec finish_gun_sec finish_net_minutes 18-39 1 1 1 Chebet, Evans Team– 1 1H 2M 20S 2H 5M 54S 2H 5M 54S M 3740 7554 7554 125.9 18-39 2 2 2 Geay, Gabriel Team– 3 1H 2M 20S 2H 6M 4S 2H 6M 4S M 3740 7564 7564 126.06666666666666 18-39 3 3 3 Kipruto, Benson Team– 5 1H 2M 19S 2H 6M 6S 2H 6M 6S M 3739 7566 7566 126.1 18-39 4 4 4 Korir, Albert Team– 19 1H 2M 20S 2H 8M 1S 2H 8M 1S M 3740 7681 7681 128.01666666666668 18-39 5 5 5 Talbi, Zouhair Team– 31 1H 2M 20S 2H 8M 35S 2H 8M 35S M 3740 7715 7715 128.58333333333334

# Camada Silver

ETL dos dfs criados. Exploração e resolução de questões encontradas

In [0]:
#Views pra ver dados com SQL

df_2023.createOrReplaceTempView("df_2023")
df_2024.createOrReplaceTempView("df_2024")

In [0]:
df_2023.columns

Out[4]: ['age_group',
 'place_overall',
 'place_gender',
 'place_division',
 'name',
 'team',
 'bib_number',
 'half_time',
 'finish_net',
 'finish_gun',
 'gender',
 'half_time_sec',
 'finish_net_sec',
 'finish_gun_sec',
 'finish_net_minutes']

In [0]:
#Explorando dados categóricos dentro dos atributos - 2023

cols_23 = ["age_group", "gender"]

for coluna in cols_23:
    print('Atributo: ',coluna)
    print(' ')
    df_2023.groupBy(coluna).count().show()

Atributo:  age_group
 
+---------+-----+
|age_group|count|
+---------+-----+
|    65-69|  770|
|    75-79|   59|
|    18-39|10168|
|      80+|    5|
|    70-74|  275|
|    50-54| 3259|
|    45-49| 4069|
|    40-44| 4119|
|    55-59| 2237|
|    60-64| 1637|
+---------+-----+

Atributo:  gender
 
+------+-----+
|gender|count|
+------+-----+
|     M|15175|
|     W|11423|
+------+-----+



In [0]:
df_2024.columns

Out[6]: ['BibNumber',
 'FullName',
 'SortName',
 'AgeOnRaceDay',
 'Gender',
 'City',
 'StateAbbrev',
 'StateName',
 'Zip',
 'CountryOfResAbbrev',
 'CountryOfResName',
 'CountryOfCtzAbbrev',
 'CountryOfCtzName',
 'OfficialTime',
 'RankOverall',
 'RankOverGender',
 'RankOverDivision',
 'EventGroup',
 'SubGroupLabel',
 'SubGroup']

In [0]:
#Explorando dados categóricos dentro dos atributos - 2024

cols_24 = ["AgeOnRaceDay", "Gender", "CountryofResAbbrev"]

for coluna in cols_24:
    print('Atributo: ',coluna)
    print(' ')
    df_2024.groupBy(coluna).count().show()

Atributo:  AgeOnRaceDay
 
+------------+-----+
|AgeOnRaceDay|count|
+------------+-----+
|          51|  671|
|          54|  491|
|          29|  539|
|          69|   82|
|          42|  709|
|          73|   34|
|          64|  200|
|          30|  507|
|          34|  469|
|          59|  323|
|          28|  585|
|          22|  170|
|          35|  724|
|          52|  613|
|          71|   60|
|          47|  758|
|          43|  716|
|          31|  503|
|          18|   25|
|          70|   93|
+------------+-----+
only showing top 20 rows

Atributo:  Gender
 
+------+-----+
|Gender|count|
+------+-----+
|     F|10904|
|     M|14578|
|     X|   48|
+------+-----+

Atributo:  CountryofResAbbrev
 
+------------------+-----+
|CountryofResAbbrev|count|
+------------------+-----+
|               POL|   87|
|               BRA|  519|
|               FRA|  205|
|               BRN|    2|
|               BOL|    1|
|               NEP|    1|
|               GIB|    1|
|               

### Principais problematicas encontradas e soluções:

######1. Atributos divergentes entre as tabelas. 
Solução: criar dataframes a partir desses apenas com os dados necessários para analise dos objetivos

######2. Formato do tempo divergente.
Comparação de tempo ficará inviavel uma vez que os formatos não batem
Solução: Transformar os formatos para segundos

######3. Juntar os dados por ano
Embora não seja necessário, facilitará o entendimentos e as analises dos objetivos
Solução: Adicionar um atributo que separe por ano (2023/2024)

######4. Formato do genero divergente
Enquanto em 2024 está com M/F (Male/Masculino, Female/Feminimo) em 2023 está com M/W (Man/Homem, Woman/Mulher). 
Solução: Onde tiver W, substituir por F para manter o padrão.

######5. Outliers de genero no df de 2024.
Existe um Genero classificado como "X"
Solução: Uma vez que não existe uma classificação dentro de um genero indefinido/não binario e como são poucos os registros (menos de 1% do total), optou-se por excluir esses registros para não prejudicar a contagem final

######6. Idade não está separada por Grupos de Idade nos dados de 2024
Solução: Criar um atributo que distribua as idades em grupos.

In [0]:
#imports necessarios

from pyspark.sql.functions import when,split, concat_ws, trim,col, regexp_extract, format_string, floor, coalesce, lit

# Soluções:

### 1. Ajustando os dataframes para tirar atributos irrelevantes

In [0]:
df_2024.columns

Out[9]: ['BibNumber',
 'FullName',
 'SortName',
 'AgeOnRaceDay',
 'Gender',
 'City',
 'StateAbbrev',
 'StateName',
 'Zip',
 'CountryOfResAbbrev',
 'CountryOfResName',
 'CountryOfCtzAbbrev',
 'CountryOfCtzName',
 'OfficialTime',
 'RankOverall',
 'RankOverGender',
 'RankOverDivision',
 'EventGroup',
 'SubGroupLabel',
 'SubGroup']

In [0]:
df_2024 = df_2024.select("FullName","BibNumber", "AgeOnRaceDay", "gender","CountryofResAbbrev","OfficialTime","RankOverall","RankOverGender")
df_2024 = df_2024.toDF("nome", "numero_peito", "idade", "genero", "país", "tempo_final","pos_final", "pos_genero")
df_2024.show(10)

+-------------+------------+-----+------+----+-----------+---------+----------+
|         nome|numero_peito|idade|genero|país|tempo_final|pos_final|pos_genero|
+-------------+------------+-----+------+----+-----------+---------+----------+
|  Sisay Lemma|           2|   33|     M| ETH|    2:06:17|        1|         1|
|  Mohamed Esa|           8|   23|     M| ETH|    2:06:58|        2|         2|
| Evans Chebet|           1|   35|     M| KEN|    2:07:22|        3|         3|
|   John Korir|           7|   27|     M| KEN|    2:07:40|        4|         4|
| Albert Korir|          14|   30|     M| KEN|    2:07:47|        5|         5|
|  Isaac Mpofu|          12|   35|     M| ZIM|    2:08:17|        6|         6|
| Cj Albertson|          25|   30|     M| USA|    2:09:53|        7|         7|
|   Yuma Morii|          35|   33|     M| JPN|    2:09:59|        8|         8|
|Cybrian Kotut|           4|   31|     M| KEN|    2:10:29|        9|         9|
|Zouhair Talbi|          11|   29|     M

In [0]:
df_2023.columns

Out[11]: ['age_group',
 'place_overall',
 'place_gender',
 'place_division',
 'name',
 'team',
 'bib_number',
 'half_time',
 'finish_net',
 'finish_gun',
 'gender',
 'half_time_sec',
 'finish_net_sec',
 'finish_gun_sec',
 'finish_net_minutes']

In [0]:
df_2023 = df_2023.select("name","bib_number", "age_group", "gender","finish_net","place_overall","place_gender")
df_2023 = df_2023.toDF("nome", "numero_peito", "faixa_etaria", "genero", "tempo_final","pos_final", "pos_genero")

df_2023 = df_2023.withColumn("nome", concat_ws(" ",
    trim(split(col("nome"), ",").getItem(1)),
    trim(split(col("nome"), ",").getItem(0))
))

df_2023.show(10)

+----------------+------------+------------+------+-----------+---------+----------+
|            nome|numero_peito|faixa_etaria|genero|tempo_final|pos_final|pos_genero|
+----------------+------------+------------+------+-----------+---------+----------+
|    Evans Chebet|           1|       18-39|     M|  2H 5M 54S|        1|         1|
|    Gabriel Geay|           3|       18-39|     M|   2H 6M 4S|        2|         2|
|  Benson Kipruto|           5|       18-39|     M|   2H 6M 6S|        3|         3|
|    Albert Korir|          19|       18-39|     M|   2H 8M 1S|        4|         4|
|   Zouhair Talbi|          31|       18-39|     M|  2H 8M 35S|        5|         5|
|  Eliud Kipchoge|           2|       18-39|     M|  2H 9M 23S|        6|         6|
|    Scott Fauble|          22|       18-39|     M|  2H 9M 44S|        7|         7|
|   Hassan Chahdi|          21|       18-39|     M|  2H 9M 46S|        8|         8|
|      John Korir|           8|       18-39|     M|  2H 10M 4S|  

### 2. Padronizando os atributos de tempo final 

In [0]:

df_2023 = df_2023.withColumn("horas", coalesce(regexp_extract("tempo_final", r"(\d+)H", 1).cast("int"), lit(0))) \
                 .withColumn("minutos", coalesce(regexp_extract("tempo_final", r"(\d+)M", 1).cast("int"), lit(0))) \
                 .withColumn("segundos", coalesce(regexp_extract("tempo_final", r"(\d+)S", 1).cast("int"), lit(0)))

# Tempo em segundos
df_2023 = df_2023.withColumn("tempo_final_segundos", 
col("horas") * 3600 + col("minutos") * 60 + col("segundos"))


df_2023 = df_2023.withColumn(
    "tempo_final",
    format_string(
        "%d:%02d:%02d",
        floor(col("tempo_final_segundos") / 3600),
        floor((col("tempo_final_segundos") % 3600) / 60),
        col("tempo_final_segundos") % 60
    )
)

df_2023 = df_2023.drop("horas", "minutos", "segundos","tempo_final_segundos")
df_2023.show(10)

+----------------+------------+------------+------+-----------+---------+----------+
|            nome|numero_peito|faixa_etaria|genero|tempo_final|pos_final|pos_genero|
+----------------+------------+------------+------+-----------+---------+----------+
|    Evans Chebet|           1|       18-39|     M|    2:05:54|        1|         1|
|    Gabriel Geay|           3|       18-39|     M|    2:06:04|        2|         2|
|  Benson Kipruto|           5|       18-39|     M|    2:06:06|        3|         3|
|    Albert Korir|          19|       18-39|     M|    2:08:01|        4|         4|
|   Zouhair Talbi|          31|       18-39|     M|    2:08:35|        5|         5|
|  Eliud Kipchoge|           2|       18-39|     M|    2:09:23|        6|         6|
|    Scott Fauble|          22|       18-39|     M|    2:09:44|        7|         7|
|   Hassan Chahdi|          21|       18-39|     M|    2:09:46|        8|         8|
|      John Korir|           8|       18-39|     M|    2:10:04|  

###3. Adicionar o ano aos dataframes


In [0]:
from pyspark.sql.functions import lit
df_2023 = df_2023.withColumn("ano", lit(2023))
df_2024 = df_2024.withColumn("ano", lit(2024))

###4. Ajustar o genero (de W pra F)

In [0]:
%sql

select * from df_2023
where genero = "W"

nome numero_peito faixa_etaria genero tempo_final pos_final pos_genero ano

In [0]:
#Views pra ver dados com SQL

df_2023.createOrReplaceTempView("df_2023")
df_2024.createOrReplaceTempView("df_2024")

In [0]:
df_2023 = df_2023.withColumn(
    "genero",
    when(col("genero") == "W", "F").otherwise(col("genero"))
)

In [0]:
df_2023.createOrReplaceTempView("df_2023")

In [0]:
%sql

select * from df_2023
where genero = "F"

nome numero_peito faixa_etaria genero tempo_final pos_final pos_genero ano Hellen Obiri 122 18-39 F 2:21:38 36 1 2023 Amane Beriso 102 18-39 F 2:21:50 40 2 2023 Lonah Salpeter 105 18-39 F 2:21:57 41 3 2023 Ababel Yeshaneh 116 18-39 F 2:22:00 42 4 2023 Emma Bates 121 18-39 F 2:22:10 44 5 2023 Nazret Weldu 114 18-39 F 2:23:25 56 6 2023 Angela Tanui 106 18-39 F 2:24:12 63 7 2023 Hiwot Gebremaryam 110 18-39 F 2:24:30 66 8 2023 Mary Ngugi 113 18-39 F 2:24:33 67 9 2023 Gotytom Gebreslase 107 18-39 F 2:24:34 68 10 2023 Aliphine Tuliamuk 129 18-39 F 2:24:37 69 11 2023 Joyciline Jepkosgei 104 18-39 F 2:24:44 70 12 2023 Viola Cheptoo 119 18-39 F 2:24:49 72 13 2023 Nell Rojas 125 18-39 F 2:24:51 74 14 2023 Nienke Brinkman 120 18-39 F 2:24:58 75 15 2023 Celestine Chepchirchir 111 18-39 F 2:25:07 76 16 2023 Desiree Linden 108 18-39 F 2:27:18 100 18 2023 Vibian Chepkirui 117 18-39 F 2:28:12 113 19 2023 Annie Frisbie 126 18-39 F 2:28:45 124 20 2023 Atsede Baysa 118 18-39 F 2:30:14 151 21 2023 Anna Rohrer 617 18-39 F 2:30:52 170 22 2023 Sydney Devore 142 18-39 F 2:31:08 178 23 2023 Maggie Montoya 130 18-39 F 2:31:19 184 24 2023 Anne-Marie Blaney 151 18-39 F 2:31:32 188 25 2023 Maegan Krifchin 131 18-39 F 2:32:46 232 26 2023 Dakotah Lindwurm 123 18-39 F 2:33:53 271 27 2023 Erika Kemp 133 18-39 F 2:33:57 273 28 2023 Savannah Berry 149 18-39 F 2:34:30 308 29 2023 Laura Thweatt 124 18-39 F 2:34:35 310 30 2023 Maurine Chepkemoi 112 18-39 F 2:35:25 348 32 2023 Militsa Mircheva 132 18-39 F 2:35:57 384 33 2023 English Tomlinson 156 18-39 F 2:36:14 397 34 2023 Kidan Kidane 7344 18-39 F 2:36:22 402 35 2023 Dani Moreno 171 18-39 F 2:38:23 519 36 2023 Mimi Smith 2050 18-39 F 2:38:33 537 37 2023 Olivia Pratt 150 18-39 F 2:38:41 547 38 2023 Liza Howard 153 18-39 F 2:40:14 644 40 2023 Tennille Ellis 161 18-39 F 2:40:37 672 41 2023 Emilee Risteen 1117 18-39 F 2:41:44 750 42 2023 Lauren Ames 1287 18-39 F 2:42:01 767 43 2023 Anne Wester 2147 18-39 F 2:42:45 837 44 2023 Melissa Paauwe 886 18-39 F 2:42:53 847 45 2023 Caroline Hentzen 1501 18-39 F 2:43:31 905 46 2023 Christine Guerrero 1630 18-39 F 2:43:50 940 47 2023 Haruka Yamaguchi 128 18-39 F 2:44:17 994 48 2023 Lizzie Mundell-Perkins 9030 18-39 F 2:44:36 1028 49 2023 Alyssa Salese 1825 18-39 F 2:44:37 1031 50 2023 Leandra Zimmermann 2958 18-39 F 2:45:04 1086 52 2023 Amanda Wiggenhorn 5518 18-39 F 2:45:38 1158 54 2023 Anna Steinman 4303 18-39 F 2:46:15 1235 57 2023 Eliza Howard 5800 18-39 F 2:46:39 1276 59 2023 Abby Rodseth 5183 18-39 F 2:46:41 1279 60 2023 Abbie Tuomi 1424 18-39 F 2:46:51 1303 61 2023 Elizabeth Camy 1167 18-39 F 2:46:51 1304 62 2023 Rachel Keane 10700 18-39 F 2:46:53 1309 63 2023 Meaghan Murray 2040 18-39 F 2:47:10 1360 64 2023 Tammy Hsieh 1534 18-39 F 2:47:17 1372 65 2023 Paula Narvaez 3028 18-39 F 2:47:27 1401 66 2023 Ewa Zaborowska 2606 18-39 F 2:48:27 1537 67 2023 Carolina Garcia 3413 18-39 F 2:48:38 1564 68 2023 Aisling Cuffe 157 18-39 F 2:48:49 1582 69 2023 Sylvia Bedford 155 18-39 F 2:48:54 1592 70 2023 Alison Wells 6022 18-39 F 2:48:55 1593 71 2023 Rachel Coogan 158 18-39 F 2:49:03 1622 73 2023 Heather Cleary 2440 18-39 F 2:49:07 1633 74 2023 Maddie Meyers 31891 18-39 F 2:49:09 1640 75 2023 Leigh Barrow 4247 18-39 F 2:49:44 1731 77 2023 Michaela Spannaus 2495 18-39 F 2:49:59 1779 79 2023 Isabella Caruso 2044 18-39 F 2:50:02 1787 80 2023 Erin Hagen 6682 18-39 F 2:50:14 1815 81 2023 Deanna Scheid 3981 18-39 F 2:51:39 2047 83 2023 Melanie Lassen 9331 18-39 F 2:51:43 2060 84 2023 Lucy Brash 3203 18-39 F 2:51:52 2086 85 2023 Jill Dennes 6912 18-39 F 2:52:13 2138 86 2023 Meghann Featherstun 2594 18-39 F 2:52:14 2145 87 2023 Ava Sorell 2289 18-39 F 2:52:21 2174 88 2023 Brooke Koblitz 7028 18-39 F 2:52:24 2182 89 2023 Caroline Lee 4142 18-39 F 2:52:32 2204 91 2023 Lainie Smith 27372 18-39 F 2:52:34 2208 92 2023 Jaime Lord 3813 18-39 F 2:52:43 2231 93 2023 Lauren Dorsky 2002 18-39 F 2:53:00 2277 95 2023 Teagan Robertson 14638 18-39 F 2:53:01 2280 96 2023 Mariah Jordan 2945 18-39 F 2:53:13 231

###5. Ajustar genero indeterminado (X)

In [0]:
#atualiza a view pra poder ver os dados atualizados
df_2024.createOrReplaceTempView("df_2024")

In [0]:
%sql

SELECT *
from df_2024
where genero = "X"

nome numero_peito idade genero país tempo_final pos_final pos_genero ano Ryan Montgomery 492 29 X USA 2:27:45 84 1 2024 Winter Parts 551 26 X USA 2:31:08 135 2 2024 Kassian Eaton 703 27 X USA 2:34:51 231 3 2024 Josh Fernandez 4513 36 X USA 2:49:28 1210 4 2024 Jacob Caswell 906 26 X USA 2:56:36 2153 5 2024 Kae Ravichandran 1019 26 X USA 2:58:16 2467 6 2024 Steven Bugarin 3297 38 X USA 2:59:46 2834 7 2024 Cal Calamia 2547 27 X USA 3:02:22 3236 8 2024 Truth Bachman 13802 33 X USA 3:08:18 4289 9 2024 Davey Mon 16394 41 X USA 3:09:05 4445 10 2024 J Solle 5297 29 X USA 3:09:30 4536 11 2024 Vinicio Vargas Carvajal 11183 52 X CRC 3:11:53 4990 12 2024 Jack Marley-Payne 12837 36 X USA 3:13:42 5394 13 2024 Alejandro Benjam�n Rodr�guez Hern�ndez 12466 29 X MEX 3:14:29 5578 14 2024 Elliot Magic Paynter 14693 41 X USA 3:17:47 6342 15 2024 Callan Knebel 14437 26 X USA 3:18:58 6648 16 2024 Goran Vendlener 9897 41 X USA 3:20:00 6903 17 2024 Teresa Howard 13130 24 X USA 3:21:27 7237 18 2024 Kylee Rood 13716 35 X USA 3:22:52 7603 19 2024 Charlie Raymer 8244 35 X USA 3:23:31 7773 20 2024 Kevin-Paul Edwards 14496 33 X USA 3:23:49 7857 21 2024 Agnes Varraine Leca 13905 39 X FRA 3:28:59 9279 22 2024 Alyssa Moskites 15130 31 X USA 3:29:49 9510 23 2024 William Paradis 17379 40 X CAN 3:30:02 9572 24 2024 Simone Pfleger 16123 39 X CAN 3:32:01 10034 25 2024 Luis Manuel Hernandez 20008 47 X MEX 3:36:30 11103 26 2024 Michael Boyd 9157 30 X USA 3:41:12 12250 27 2024 Oscar Villalpando 19060 44 X MEX 3:41:43 12363 28 2024 Nonie Andersen 11728 27 X USA 3:44:18 12965 29 2024 Caleb-Michael Files 30437 32 X USA 3:45:48 13312 30 2024 Darwin Romero 11363 39 X USA 3:46:50 13545 31 2024 Nuttanai Anuntarumporn 10447 36 X THA 3:49:30 14148 32 2024 Carlos Hilario Navas Montilla 6994 42 X ESP 3:50:53 14432 33 2024 Nenad Moconja 17974 40 X USA 3:54:35 15207 34 2024 Jessy Parker Humphreys 16507 29 X GBR 3:55:09 15320 35 2024 Mario Cesar Santos 11140 35 X BRA 3:59:45 16178 36 2024 Rob Roberts 18822 46 X USA 4:01:15 16416 37 2024 Andrew Holderness 23189 62 X GBR 4:03:53 16823 38 2024 Danielle Bishop 17213 29 X USA 4:05:58 17123 39 2024 Adam Marsh 16676 51 X USA 4:26:00 19652 40 2024 Robert Fortuin 22313 58 X USA 4:27:20 19795 41 2024 Cori Harral 13119 27 X USA 4:32:28 20332 42 2024 Marcus Carmicle 19622 53 X USA 4:35:39 20632 43 2024 Sam Cote 28787 23 X USA 4:35:47 20646 44 2024 Ian Chubb 24681 55 X GBR 4:36:50 20734 45 2024 Glenique Frank 30078 54 X GBR 4:58:49 22666 46 2024 Shay Meyer 25505 43 X USA 5:11:46 23418 47 2024 Christian Rountry 25411 55 X USA 6:06:49 25447 48 2024

In [0]:
df_2024 = df_2024.filter(col("genero") != "X")
df_2024.createOrReplaceTempView("df_2024")

In [0]:
%sql

select * from df_2024 where genero = "X"

nome numero_peito idade genero país tempo_final pos_final pos_genero ano

###6. Ajustar grupos de idade

In [0]:
df_2024 = df_2024.withColumn(
    "faixa_etaria",
    when(col("idade") < 18, "Under 18")
    .when((col("idade") >= 18) & (col("idade") <= 39), "18-39")
    .when((col("idade") >= 40) & (col("idade") <= 44), "40-44")
    .when((col("idade") >= 45) & (col("idade") <= 49), "45-49")
    .when((col("idade") >= 50) & (col("idade") <= 54), "50-54")
    .when((col("idade") >= 55) & (col("idade") <= 59), "55-59")
    .when((col("idade") >= 60) & (col("idade") <= 64), "60-64")
    .when((col("idade") >= 65) & (col("idade") <= 69), "65-69")
    .when((col("idade") >= 70) & (col("idade") <= 74), "70-74")
    .when((col("idade") >= 75) & (col("idade") <= 79), "75-79")
    .when(col("idade") >= 80, "80+")
    .otherwise("INVALIDO")
)

In [0]:
df_2024 = df_2024.drop ("idade")

In [0]:
df_2024.show(10)

+-------------+------------+------+----+-----------+---------+----------+----+------------+
|         nome|numero_peito|genero|país|tempo_final|pos_final|pos_genero| ano|faixa_etaria|
+-------------+------------+------+----+-----------+---------+----------+----+------------+
|  Sisay Lemma|           2|     M| ETH|    2:06:17|        1|         1|2024|       18-39|
|  Mohamed Esa|           8|     M| ETH|    2:06:58|        2|         2|2024|       18-39|
| Evans Chebet|           1|     M| KEN|    2:07:22|        3|         3|2024|       18-39|
|   John Korir|           7|     M| KEN|    2:07:40|        4|         4|2024|       18-39|
| Albert Korir|          14|     M| KEN|    2:07:47|        5|         5|2024|       18-39|
|  Isaac Mpofu|          12|     M| ZIM|    2:08:17|        6|         6|2024|       18-39|
| Cj Albertson|          25|     M| USA|    2:09:53|        7|         7|2024|       18-39|
|   Yuma Morii|          35|     M| JPN|    2:09:59|        8|         8|2024|  

In [0]:
%sql

select * from df_2024 where idade = "INVALIDO"

nome numero_peito idade genero país tempo_final pos_final pos_genero ano

# Tabelas Fato e Dimensão

In [0]:
#Separar as colunas em fato e dim

from pyspark.sql.functions import monotonically_increasing_id

In [0]:
df_maratona = df_2024.unionByName(df_2023, allowMissingColumns=True)
df_maratona.show()

+--------------------+------------+------+----+-----------+---------+----------+----+------------+
|                nome|numero_peito|genero|país|tempo_final|pos_final|pos_genero| ano|faixa_etaria|
+--------------------+------------+------+----+-----------+---------+----------+----+------------+
|         Sisay Lemma|           2|     M| ETH|    2:06:17|        1|         1|2024|       18-39|
|         Mohamed Esa|           8|     M| ETH|    2:06:58|        2|         2|2024|       18-39|
|        Evans Chebet|           1|     M| KEN|    2:07:22|        3|         3|2024|       18-39|
|          John Korir|           7|     M| KEN|    2:07:40|        4|         4|2024|       18-39|
|        Albert Korir|          14|     M| KEN|    2:07:47|        5|         5|2024|       18-39|
|         Isaac Mpofu|          12|     M| ZIM|    2:08:17|        6|         6|2024|       18-39|
|        Cj Albertson|          25|     M| USA|    2:09:53|        7|         7|2024|       18-39|
|         

In [0]:
#dim participante

dim_participante = df_maratona.select("nome", "numero_peito", "tempo_final").distinct()

# id participante
dim_participante = dim_participante.withColumn("id_participante", monotonically_increasing_id())
dim_participante.show(5)

+----------------+------------+-----------+---------------+
|            nome|numero_peito|tempo_final|id_participante|
+----------------+------------+-----------+---------------+
|  Joseph Purtell|         365|    2:34:31|              0|
|    Joseph Craig|        1396|    2:41:54|              1|
|    Jeremy Bauer|         970|    2:47:06|              2|
|Francisco Medina|        1315|    2:48:28|              3|
|     Aaron Roche|        2721|    2:50:12|              4|
+----------------+------------+-----------+---------------+
only showing top 5 rows



In [0]:
#dim categoria

dim_categoria = df_maratona.select("genero", "faixa_etaria").distinct()
dim_categoria = dim_categoria.withColumn("id_categoria", monotonically_increasing_id())

dim_categoria.show(5)

+------+------------+------------+
|genero|faixa_etaria|id_categoria|
+------+------------+------------+
|     M|       55-59|           0|
|     F|       45-49|           1|
|     M|       18-39|           2|
|     M|       60-64|           3|
|     M|       40-44|           4|
+------+------------+------------+
only showing top 5 rows



In [0]:
#fato maratona

df_fato = df_maratona.join(dim_participante, ["nome", "numero_peito", "tempo_final"], "left")
df_fato = df_fato.join(dim_categoria, ["genero", "faixa_etaria"], "left")

fato_maratona = df_fato.select(
    monotonically_increasing_id().alias("id_fato"),
    "ano",
    "id_participante",
    "id_categoria",
    "pos_final",
    "pos_genero"
)

fato_maratona.show(5)

+-------+----+---------------+------------+---------+----------+
|id_fato| ano|id_participante|id_categoria|pos_final|pos_genero|
+-------+----+---------------+------------+---------+----------+
|      0|2024|          14404|           2|        1|         1|
|      1|2024|          19395|           2|        2|         2|
|      2|2024|          19074|           2|        3|         3|
|      3|2024|           5354|           2|        4|         4|
|      4|2024|          18296|           2|        5|         5|
+-------+----+---------------+------------+---------+----------+
only showing top 5 rows



#Camada Gold
##Analise dos Objetivos

###Objetivos:

- Comparar o numero de inscritos entre os anos de 2023 e 2024 na Maratona de Boston para entender se realmente houve um aumento no volume de inscrições;
- Entender se, comparativamente, houve diferença na porcentagem dos inscritos por genero;
- Entender se, comparativamente, houve diferença nos inscritos por faixa etaria;
- Analisar a media de tempos de finalização em grupos de genero para entender se houve alguma diferença entre os anos;
- Analisar a media de tempos em cada grupo por faixa etaria e genero para entender se houve melhora ou piora entre anos;
- ~~Analisar se houve diversidade da nacionalidade dos participantes durante os anos~~ (Objetivo não pode ser medido)

In [0]:
#Views da fato e dims pra resolver problemad com SQL

fato_maratona.createOrReplaceTempView("fato_maratona")
dim_participante.createOrReplaceTempView("dim_participante")
dim_categoria.createOrReplaceTempView("dim_categoria")

### Objetivo 1:

Comparar o numero de inscritos entre os anos de 2023 e 2024 na Maratona de Boston para entender se realmente houve um aumento no volume de inscrições

In [0]:
%sql
select ano, count(*) as total_participantes from fato_maratona
group by ano

ano total_participantes 2024 25482 2023 26598

###Nota sobre objetivo 1:
Honestamente, esse resultado foi surpreendente. Esperava um aumento nas inscrições. Porém isso pode ser explicado pelo fato de que a maratona de Boston exige um indice (tempo maximo que um atleta tem que ter, comprovadamente,concluido uma maratona, documentado em uma prova oficial) para entrada sem sorteio.

Em 2023, esse indice, para faixa etaria de 18 a 39 anos, era de 3 horas. Em 2024, foi de 2h52minutos, exigindo que para se qualificar, os atletas provassem que correm 1km em aproximadamente 4 minutos. A diferença parece pequena, mas em longos percursos, é uma alta velocidade.

Embora existam outros meios, os mesmos são mais custosos financeiramente. Portanto, pode-se esperar que esse seja um fator que fez diferença nas inscrições.

### Objetivo 2:
Entender se, comparativamente, houve diferença na porcentagem dos inscritos por genero em cada ano

In [0]:
%sql
  SELECT f.ano,d.genero, COUNT(*) AS total
  FROM fato_maratona f
  JOIN dim_categoria d ON f.id_categoria = d.id_categoria
  GROUP BY f.ano, d.genero
  order by d.genero, f.ano

ano genero total 2023 F 11423 2024 F 10904 2023 M 15175 2024 M 14578

###Nota sobre objetivo 2:
Como esperado pela diminuição de inscrições, em ambos os generos o numero de participantes caiu.
Comparado com 2023, 2024 teve uma diminuição de 4.6% de participantes mulheres e 4% de participantes homens. Porém, comparativamente, houve um ligeiro aumento na porcentagem de mulheres inscritas em 2024, indo de 42.7% no ano de 2023 para 43% no ano de 2024. Pode-se concluir então que mesmo com uma tendencia a diminuir as incrições, houve uma igualdade ligeiramente maior na questão de genero no ano de 2024.

### Objetivo 3
•	Entender se, comparativamente, houve diferença nos inscritos por faixa etaria

In [0]:
%sql
SELECT f.ano,d.faixa_etaria,COUNT(*) AS total
  FROM fato_maratona f
  JOIN dim_categoria d ON f.id_categoria = d.id_categoria
  GROUP BY f.ano, d.faixa_etaria
  ORDER BY d.faixa_etaria, f.ano

ano faixa_etaria total 2023 18-39 10168 2024 18-39 9498 2023 40-44 4119 2024 40-44 3615 2023 45-49 4069 2024 45-49 3972 2023 50-54 3259 2024 50-54 3121 2023 55-59 2237 2024 55-59 2404 2023 60-64 1637 2024 60-64 1723 2023 65-69 770 2024 65-69 833 2023 70-74 275 2024 70-74 255 2023 75-79 59 2024 75-79 50 2023 80+ 5 2024 80+ 11

###Nota sobre objetivo 3:
De forma parecida com o objetivo 2, há alguns pontos interessantes a serem observados.

Até os 54 anos, todas as faixas tiveram quedas consideraveis, porém, a partir dos 55, acontece um fato interessante e o movimento é o contrario: todas as inscrições subiram.  A faixa etaria de 80+, em especial, teve um aumento de 100% sua participação, indo de 5 para 11 pessoas

Isso ainda pode ser explicado pelo índice exigido. Como a cada faixa etaria esse tempo aumenta, vê-se que nas faixas mais jovens é mais inalcansavel do que para os mais velhos.


###Objetivo 4:
Analisar a media de tempos de finalização em grupos de genero para entender se houve alguma diferença entre os anos

In [0]:
#transformar str de tempo final em segundos para poder fazer contas (já devia ter feito isso la em cima mas não me toquei que eram strings)

dim_participante = dim_participante.withColumn("hora", split(col("tempo_final"), ":")[0].cast("int")) \
                                   .withColumn("minuto", split(col("tempo_final"), ":")[1].cast("int")) \
                                   .withColumn("segundo", split(col("tempo_final"), ":")[2].cast("int"))


dim_participante = dim_participante.withColumn(
    "tempo_segundos",
    col("hora") * 3600 + col("minuto") * 60 + col("segundo")
)


dim_participante = dim_participante.drop("hora", "minuto", "segundo")
dim_participante.select("tempo_final", "tempo_segundos").show(10)


+-----------+--------------+
|tempo_final|tempo_segundos|
+-----------+--------------+
|    2:10:29|          7829|
|    2:09:59|          7799|
|    2:07:40|          7660|
|    2:06:17|          7577|
|    2:08:17|          7697|
|    2:10:45|          7845|
|    2:10:52|          7852|
|    2:07:47|          7667|
|    2:07:22|          7642|
|    2:06:58|          7618|
+-----------+--------------+
only showing top 10 rows



In [0]:
dim_participante.show()

+--------------------+------------+-----------+---------------+--------------+
|                nome|numero_peito|tempo_final|id_participante|tempo_segundos|
+--------------------+------------+-----------+---------------+--------------+
|      Joseph Purtell|         365|    2:34:31|              0|          9271|
|        Joseph Craig|        1396|    2:41:54|              1|          9714|
|        Jeremy Bauer|         970|    2:47:06|              2|         10026|
|    Francisco Medina|        1315|    2:48:28|              3|         10108|
|         Aaron Roche|        2721|    2:50:12|              4|         10212|
|        Siarhei Trus|        1916|    2:51:25|              5|         10285|
|  Christian Brensing|        5684|    2:52:34|              6|         10354|
|Harshavardhan Har...|        4803|    2:54:16|              7|         10456|
|      Daniel Moncada|        1673|    2:54:47|              8|         10487|
|    Niklas Vom Felde|        4520|    2:54:54|     

In [0]:
dim_categoria.createOrReplaceTempView("dim_categoria")
dim_participante.createOrReplaceTempView("dim_participante")

In [0]:
%sql
SELECT f.ano, cat.genero, ROUND(AVG(part.tempo_segundos) / 60,2) AS media_tempo_minutos
FROM fato_maratona f
JOIN dim_participante part ON f.id_participante = part.id_participante
JOIN dim_categoria cat ON f.id_categoria = cat.id_categoria
GROUP BY f.ano, cat.genero
ORDER BY cat.genero, f.ano

ano genero media_tempo_minutos 2023 F 235.93 2024 F 245.29 2023 M 212.43 2024 M 223.99

###Nota sobre objetivo 4:

No quesito da velocidade media, embora ambos tenham tido uma caída no tempo médio, o genero feminino manteve mais a constancia em velocidade, apenas caindo 3.96% comparativamente com o ano de 2023. Já o genero Masculino teve uma queda maior, diminuindo em 5.44% a média vinda do ano anterior

### Objetivo 5
Analisar a media de tempos em cada grupo por faixa etaria e genero para entender se houve melhora ou piora entre anos

In [0]:
%sql
SELECT f.ano,cat.genero, cat.faixa_etaria, ROUND(AVG(part.tempo_segundos) / 60, 2) AS media_tempo_minutos
FROM fato_maratona f
JOIN dim_participante part ON f.id_participante = part.id_participante
JOIN dim_categoria cat ON f.id_categoria = cat.id_categoria
GROUP BY f.ano, cat.genero, cat.faixa_etaria
ORDER BY cat.genero, cat.faixa_etaria, f.ano

ano genero faixa_etaria media_tempo_minutos 2023 F 18-39 227.91 2024 F 18-39 236.06 2023 F 40-44 230.74 2024 F 40-44 238.55 2023 F 45-49 235.87 2024 F 45-49 242.74 2023 F 50-54 244.61 2024 F 50-54 255.45 2023 F 55-59 253.28 2024 F 55-59 264.11 2023 F 60-64 258.9 2024 F 60-64 274.27 2023 F 65-69 272.1 2024 F 65-69 290.73 2023 F 70-74 293.4 2024 F 70-74 298.81 2023 F 75-79 296.37 2024 F 75-79 339.16 2023 F 80+ 305.9 2024 F 80+ 354.52 2023 M 18-39 196.51 2024 M 18-39 206.86 2023 M 40-44 203.0 2024 M 40-44 211.53 2023 M 45-49 210.26 2024 M 45-49 221.88 2023 M 50-54 221.32 2024 M 50-54 230.95 2023 M 55-59 228.29 2024 M 55-59 238.94 2023 M 60-64 239.81 2024 M 60-64 252.23 2023 M 65-69 252.95 2024 M 65-69 267.03 2023 M 70-74 270.18 2024 M 70-74 288.71 2023 M 75-79 296.32 2024 M 75-79 300.61 2023 M 80+ 297.91 2024 M 80+ 297.0

###Nota sobre objetivo 5:

Novamente, entre os aspectos, existem observações importantes de se fazer.
No geral, a porcentagem da media de tempo caiu pouco (entre 3 e 5% a mais que 2023) para o genero feminino nas categorias mais jovens, até seus 55 anos. Porém nas categorias mais idosas, a porcentagem aumentou muito, chegando aos 15% nos 80+. O contrario se observou nas categorias masculinas, onde os mais jovens tiveram uma variação mais alta (entre 5 e 7% a mais que em 2023), porém observando as idades mais avançadas, o crescimento é minimo (1,8% apenas na faixa entre 75 e 79 anos em comparação com 2023) e até chega, na categoria masculina 80+, a ser 0.31% mais rapida que a media de 2023

###Nota sobre objetivo 6:

Ao visualizar o primeiro dataframe (2024), o objetivo de analisar os atletas por sua nacionalidade e até mesmo comparar qual a porcentagem de atletas vindos de outros países e continentes virou uma ideia. Porém não foi até chegar na limpeza que finalmente me dei conta que não seria possivel, uma vez que a tabela de base 2023 não possui os dados de nacionalidade dos participantes. Talvez para um próximo projeto, isso seria algo a ser buscado, mas no momento não foi possivel separar as nacionalidades 